HIVE-27972: Set 'tez' as default value in hive.execution.engine

abstractdog opened the pull request:
### What changes were proposed in this pull request?
Changing the default value of hive.execution.engine to "tez" in HiveConf and handle all the mess that comes after.

this patch contains lots of changes according to a typical pattern of points below:
1. basic change in HiveConf
2. introducing HiveConfForTest, which is a convenience HiveConf for tests, which takes care of setting up proper Tez values without having any Junit magic.
3. using HiveConfForTest where it's applicable and needed
4. added TODO comments where it was too hard to adapt the test to Tez (long-long replication tests) or it doesn't make sense or isn't worth (e.g. hcatalog)
5. copied the default hive-site.xml from  data/conf/ to data/conf/mr, denoting that from now on, "mr" is a special value (because it's not the default anymore) - data/conf/hive-site.xml remains unchanged which means Tez there from now on
6. adapted some qtest drivers to change in 5), so using the new, "mr" related conf
7. remove all occurrences of hive.execution.engine=tez from site xmls, and commented where it's still present, double-checked by:
```
mvn clean -Pitests,iceberg
 grep -iRH "hive.execution.engine" --include="*.xml" --exclude-dir=target -C 2
```
8. added test scope dependencies where it was needed, not a surprise: unit tests ran on MR until now, so they had hadoop mapreduce as dependency, not they must have tez
9. code formatting fixes where it really bothered me (e.g. unused imports)
10. miscellaneous fixed in unit tests, which came from the differences between MR and Tez (and Hive's integration of them)

### Why are the changes needed?
Tez is the supported engine, and needs to be reflected in code.

### Does this PR introduce _any_ user-facing change?
No. Depending on their config, customers running on real clusters already use Tez (or MR, sigh). HiveConf default affects code/UT mostly.

### Is the change a dependency upgrade?
No upgrade, only tez dependencies were brought into module where it was needed.

### How was this patch tested?
Unit tests are all passing. Where I found challenges, I created follow-up tickets.

akshat0395 commented on common/src/java/org/apache/hadoop/hive/conf/HiveConf.java:
```diff
@@ -4564,7 +4564,7 @@ public static enum ConfVars {
 
-    HIVE_EXECUTION_ENGINE("hive.execution.engine", "mr", new StringSet(true, "mr", "tez"),
+    HIVE_EXECUTION_ENGINE("hive.execution.engine", "tez", new StringSet(true, "tez", "mr"),
         "Chooses execution engine. Options are: mr (Map reduce, default), tez. While MR\n" +
```
Can we update the text as well,
"Chooses execution engine. Options are: tez (default), mr(Map reduce). MR\n" +
        "is a historical engine\n" +
        "and is deprecated in Hive 2 line. It may be removed without further warning."),

abstractdog replied:
sure, this first commit is just for a quick precommit sanity test

akshat0395 replied:
Make sense, thx

abstractdog replied:
this has been addressed in recent commits

InvisibleProgrammer commented:
Hi, there are some possibly related stuff that came into my mind: 

Do we need to change some configuration files as well? 
My first guess is data/conf/hive-site.xml. It can overwrite the default values. 
And also, I'm thinking about the docker stuff as well. 
And StartMiniHS2Cluster overwrites it as well. 

The other thing that I don't know for sure is what is the default engine for qtests?

abstractdog commented:
> Hi, there are some possibly related stuff that came into my mind:
> 
> Do we need to change some configuration files as well? My first guess is data/conf/hive-site.xml. It can overwrite the default values. And also, I'm thinking about the docker stuff as well. And StartMiniHS2Cluster overwrites it as well.
> 
> The other thing that I don't know for sure is what is the default engine for qtests?

regarding config files: yes I guess, haven't checked though

regarding qtests: default engine for qtests depends on the actual driver, those have to be handled one-by-one:
https://github.com/apache/hive/blob/master/itests/util/src/main/java/org/apache/hadoop/hive/ql/QTestMiniClusters.java

zratkai commented on common/src/java/org/apache/hadoop/hive/conf/HiveConfForTest.java:
```diff
@@ -0,0 +1,100 @@
+ * This class is mostly used in unit test environment to prevent polluting the
+ * original HiveConf, and also to provide a way to set some default values for execution.
+ */
+public class HiveConfForTest extends HiveConf {
```
As it is used only for tests, I think it should be in src/test/java folder.

abstractdog replied:
yes, I was thinking about this, maybe it needs some pom.xml changes to refer to it, let me check

abstractdog replied:
done

abstractdog commented on common/src/java/org/apache/hadoop/hive/conf/HiveConfForTest.java:
```diff
@@ -0,0 +1,100 @@
+   */
+  public String getOverlayOptionsAsQueryString() {
+    String queryString = Joiner.on(";").withKeyValueSeparator("=").join(overlay);
+    System.out.println("QUERY STRING: " + queryString);
```
oops, leftover, I'll remove

abstractdog replied:
removed

aturoczy reviewed (approved):
[I want it all](https://www.youtube.com/watch?v=hFDcoX7s6rE)

SourabhBadhya commented on service/pom.xml:
```diff
@@ -319,6 +319,11 @@
         </exclusion>
       </exclusions>
     </dependency>
+    <dependency>
+      <groupId>org.apache.tez</groupId>
+      <artifactId>tez-mapreduce</artifactId>
+      <scope>test</scope>
+    </dependency>
```
Why is this required?

abstractdog replied:
otherwise test will fail with MRInput class not found or something similar

abstractdog replied:
checked, now all the affected unit tests seem to pass without this dependency, so removing from this PR (maybe a UT which needed this was finally reverted to MR due to some issues and will be taken care of by one of the followup tickets)

SourabhBadhya commented on streaming/pom.xml:
```diff
@@ -126,19 +126,43 @@
       <version>${project.version}</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.apache.hive</groupId>
+      <artifactId>hive-exec</artifactId>
+      <version>${project.version}</version>
+      <type>test-jar</type>
+      <scope>test</scope>
+    </dependency>
     <dependency>
       <groupId>org.apache.hive</groupId>
       <artifactId>hive-standalone-metastore-server</artifactId>
       <version>4.1.0-SNAPSHOT</version>
       <classifier>tests</classifier>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.apache.hive</groupId>
+      <artifactId>hive-common</artifactId>
+      <version>${project.version}</version>
+      <classifier>tests</classifier>
+      <scope>test</scope>
+    </dependency>
     <dependency>
       <groupId>org.apache.hadoop</groupId>
       <artifactId>hadoop-mapreduce-client-jobclient</artifactId>
       <version>${hadoop.version}</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.apache.tez</groupId>
+      <artifactId>tez-mapreduce</artifactId>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.tez</groupId>
+      <artifactId>tez-dag</artifactId>
+      <scope>test</scope>
+    </dependency>
```
Why is this required?

abstractdog replied:
tez-mapreduce: MRInput (as mentioned above)
tez-dag: I cannot recall I'll double-check
hive-exec: I cannot recall I'll double-check
hive-common: HiveConfForTest

abstractdog replied:
without tez-mapreduce it fails, it's needed
```
java.lang.NoClassDefFoundError: org/apache/tez/mapreduce/hadoop/InputSplitInfo
	at org.apache.hadoop.hive.ql.plan.TezWork.<init>(TezWork.java:91)
	at org.apache.hadoop.hive.ql.parse.GenTezProcContext.<init>(GenTezProcContext.java:176)
	at org.apache.hadoop.hive.ql.parse.TezCompiler.generateTaskTree(TezCompiler.java:634)
	at org.apache.hadoop.hive.ql.parse.TaskCompiler.compile(TaskCompiler.java:302)
...
Caused by: java.lang.ClassNotFoundException: org.apache.tez.mapreduce.hadoop.InputSplitInfo
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	... 67 more
```

without tez-dag:
```
Caused by: java.lang.ClassNotFoundException: org.apache.tez.dag.app.TezTaskCommunicatorImpl
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382) ~[?:1.8.0_292]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418) ~[?:1.8.0_292]
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:352) ~[?:1.8.0_292]
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351) ~[?:1.8.0_292]
	... 83 more
```

hive-exec: removed, it was a leftover from early versions of this patch when HiveConfForTest was in the hive-exec module

hive-common: needed because of HiveConfForTest
without it:
```[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M4:test (default-test) on project hive-streaming: Execution default-test of goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M4:test failed: org.junit.platform.commons.JUnitException: TestEngine with ID 'junit-jupiter' failed to discover tests: ClassSelector [className = 'org.apache.hive.streaming.TestStreamingDynamicPartitioning'] resolution failed: org/apache/hadoop/hive/conf/HiveConfForTest: org.apache.hadoop.hive.conf.HiveConfForTest -> [Help 1]
```

qris replied:
I just found this commit because I'm experimenting with Hive and I couldn't insert any data into a new instance because tez is now the default execution engine, but Hive doesn't ship with the required jars and the installation instructions make no mention of the need to install them, or how, so I get this same error message:
```
Caused by: java.lang.NoClassDefFoundError: org/apache/tez/mapreduce/hadoop/InputSplitInfo
```
In the end I followed the instructions [here](https://hive.apache.org/docs/latest/manual-installation_283118363/#tez) but using a tez binary distribution tarball instead of compiling it, which seems to partly work but Hive seems to be stuck on "INFO  : Tez session hasn't been created yet. Opening session", so I'm reverting to mr. Please could someone update the installation instructions?

abstractdog commented on itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestAcidOnTez.java:
```diff
@@ -111,7 +113,11 @@ public String toString() {
-    hiveConf = new HiveConf(this.getClass());
+    hiveConf = new HiveConfForTest(getClass());
+
+    hiveConf.set("hive.tez.container.size", "128");
+    hiveConf.setBoolean("hive.merge.tezfiles", false);
```
it was already part of this test class, I didn't want to touch them
I'll change them to ConfVars

abstractdog replied:
done

abstractdog commented on itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestAcidOnTez.java:
```diff
@@ -1002,22 +984,6 @@ private void runQueries(String engine, String joinType, HiveConf confForTez, Hiv
     }
   }
 
-  public static void setupTez(HiveConf conf) {
-    conf.setVar(HiveConf.ConfVars.HIVE_EXECUTION_ENGINE, "tez");
-    conf.setVar(HiveConf.ConfVars.HIVE_USER_INSTALL_DIR, TEST_DATA_DIR);
-    conf.set("tez.am.resource.memory.mb", "128");
-    conf.set("tez.am.dag.scheduler.class", "org.apache.tez.dag.app.dag.impl.DAGSchedulerNaturalOrderControlled");
-    conf.setBoolean("tez.local.mode", true);
-    conf.setBoolean("tez.local.mode.without.network", true);
-    conf.set("fs.defaultFS", "file:///");
-    conf.setBoolean("tez.runtime.optimize.local.fetch", true);
-    conf.set("tez.staging-dir", TEST_DATA_DIR);
-    conf.setBoolean("tez.ignore.lib.uris", true);
-    conf.set("hive.tez.container.size", "128");
-    conf.setBoolean("hive.merge.tezfiles", false); 
-    conf.setBoolean("hive.in.tez.test", true);
-  }
```
yeah! some of the opts are now taken care of there

abstractdog commented on itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/TestAcidOnTez.java:
```diff
@@ -57,6 +58,7 @@
 import org.apache.hadoop.hive.ql.io.orc.OrcInputFormat;
 import org.apache.hadoop.hive.ql.io.orc.Reader;
 import org.apache.hadoop.hive.ql.session.SessionState;
+
```
will remove

abstractdog replied:
done

abstractdog commented on hcatalog/pom.xml:
```diff
@@ -79,6 +79,13 @@
       <classifier>tests</classifier>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.apache.hive</groupId>
+      <artifactId>hive-common</artifactId>
+      <version>${project.version}</version>
+      <classifier>tests</classifier>
+      <scope>test</scope>
+    </dependency>
```
HiveConfForTest usage
I'm checking if any of the catalog tests were adapted to tez, if so, this is needed, if it isn't yet, I'll remove

abstractdog replied:
checked, needed for TestHCatAuthUtil for example
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M4:test (default-test) on project hive-hcatalog-core: Execution default-test of goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M4:test failed: java.lang.NoClassDefFoundError: org/apache/hadoop/hive/conf/HiveConfForTest: org.apache.hadoop.hive.conf.HiveConfForTest -> [Help 1]
```

abstractdog commented:
> I have complied hive with this PR, built a docker image using docker compose with postgres and tested out a few queries, everything worked fine.

thanks a lot!!

abstractdog commented on pom.xml:
```diff
@@ -1334,11 +1334,36 @@
             <groupId>io.netty</groupId>
             <artifactId>netty</artifactId>
           </exclusion>
+          <exclusion>
+            <groupId>org.apache.hadoop</groupId>
+            <artifactId>hadoop-common</artifactId>
+          </exclusion>
+          <exclusion>
+            <groupId>org.apache.hadoop</groupId>
+            <artifactId>hadoop-mapreduce-client-core</artifactId>
+          </exclusion>
+          <exclusion>
+            <groupId>org.apache.hadoop</groupId>
+            <artifactId>hadoop-mapreduce-client-jobclient</artifactId>
+          </exclusion>
+          <exclusion>
+            <groupId>org.apache.hadoop</groupId>
+            <artifactId>hadoop-mapreduce-client-common</artifactId>
+          </exclusion>
+          <exclusion>
+            <groupId>org.apache.hadoop</groupId>
+            <artifactId>hadoop-hdfs</artifactId>
+          </exclusion>
+          <exclusion>
+            <groupId>org.apache.hadoop</groupId>
+            <artifactId>hadoop-yarn-client</artifactId>
+          </exclusion>
```
this is added to the dependencyManagement section, so from now on, every <dependency> in sub-modules will exclude the same Hadoop transitive dependencies
the reason I added these exclusions was to make it inline with other tez dependencies defined here as a best effort (assuming that exclusions were added for a reason earlier):
https://github.com/apache/hive/blob/1c9969a003b09abc851ae7e19631ad208d3b6066/pom.xml#L1344-L1387

abstractdog commented:
thanks a lot, @SourabhBadhya  for the review, good catches, I addressed all your comments in 
https://github.com/apache/hive/pull/4973/commits/59ad09bc8d1b91514953f07fd5207d3387d36815

simhadri-g reviewed (approved):
LGTM +1 (pending tests)

abstractdog commented:
> LGTM +1 (pending tests)

thanks @simhadri-g
let me wait for @SourabhBadhya if he has any comments, objections

SourabhBadhya reviewed (approved):
LGTM +1 (pending tests)
